### PR TITLE
datetime: use numerical values in db

### PR DESCRIPTION
**Pull request by phweyland**
Replace the exif datetime format in db (and img cache) by a numerical value.
There are a lot of impacted files. I'm progressing function by function to try to keep control.
Comments are welcome.

### CODE REVIEW

**Comment by AlicVB** [comment]
Great work. This will simplify lot of things for parts of code that deals with dates !
Thanks for doing that tedious job !

**Comment by phweyland** [comment]
I've done what I wanted to do:
- datetime_taken and timestamp based on GTimeSpan (numerical value) relative to 0001:01:01 00:00:00.
- make work again metadata_view/collect/collection/timeline, xmp/import recognition and gpx/geotagging.

It would be great if some good soul(s) could shake a bit the thing ...

@TurboGit I see you have set the PR as draft. What's the difference with the wip label ?

**Comment by TurboGit** [comment]
@phweyland : A draft PR cannot be merged (button disabled) so it avoids me messing with a PR :) A safety belt if you prefer.

**Comment by wpferguson** [comment]
I'm not sure what this is going to do to lua, and how many scripts will be affected.  I'll try and set up a test environment and check it out, but I'm super busy right now.

**Comment by phweyland** [comment]
> I'm not sure what this is going to do to lua

I haven't checked specifically on this side, but if lua doesn't access directly (I haven't found anything) to img cache or database datetime(s), it should be safe. If not, that means there are some bugs in datetime.c.
time_t (unix time) shoule be used only with file related timestamp.

@TurboGit I've seen some reference to `export_timestamp = some values` in test files: these values are to be wrong now (GTimeStamp instead of time_t)

**Comment by paolobenve** [comment]
I'm seeing that the variable `datetime_taken` is changed from a character representation to an integere.

I'm wondering if it wouldn't be better to change the variable name.

I fear that keeping the same name may confuse anyone who examines the code in the future.

**Comment by wpferguson** [comment]
>  haven't checked specifically on this side, but if lua doesn't access directly

lua does access it directly, dt_lua_image_t.exif_datetime_taken (https://docs.darktable.org/lua/stable/lua.api.manual/types/dt_lua_image_t/).

There are 8 scripts in the lua-scripts repository that use it, or use it and change it.

**Comment by phweyland** [comment]
I see, `exif_datetime_taken` is accessed from/to the image cache.
For the lua script, `dt_lua_image_t.exif_datetime_taken` is a read/write attribute, right ?
In lua language, attribute or method are equivalent, aren't they ?
If yes, there should be a way to call dt_datetime_img_to_exif() and dt_datetime_exif_to_img()  instead of reading/writing the datetime string from/to the cache. That would avoid to jeopardize the scripts ...

**Comment by wpferguson** [comment]
> For the lua script, dt_lua_image_t.exif_datetime_taken is a read/write attribute, right ?

Yes

> In lua language, attribute or method are equivalent, aren't they ?

Most of the time, but not always

> If yes, there should be a way to call dt_datetime_img_to_exif() and dt_datetime_exif_to_img() instead of reading/writing the datetime string from/to the cache (EDIT: like for dt_lua_image_t.path for example). That would avoid to jeopardize the scripts ...

I'll look at it and see...

**Comment by wpferguson** [comment]
look at src/lua/image.c, line 474 `dt_lua_init_image()`.  A datatype is created then automagically populated from the data (i.e. fields are mapped to the data).

**Comment by phweyland** [comment]
> I'm seeing that the variable `datetime_taken` is changed from a character representation to an integere.
> I'm wondering if it wouldn't be better to change the variable name.

The name itself doesn't tell anything about the variable type, so I I don't think that's misleading. Of course somebody who expects a string may miss the new type. But the compiler will cry, won't it ?
Anyway, that's easy to change, if other people think we should.

**Comment by wpferguson** [comment]
@phweyland Thanks, I'll test it tonight

**Comment by wpferguson** [comment]
@phweyland Tried a script and it got a blank for the exif_datetime_taken field.  Let me play with it a little bit and see what I can come up with.

**Comment by phweyland** [comment]
On my side I've used the scripts `AutoGrouper` which provides `group gap [sec.]` parameter which read datetime, and `image_time` which set datetime too. Both work as expected.

**Comment by phweyland** [comment]
@wpferguson, I've fixed the case when an image has no date. It now return "" instead the wrongly formatted origin date.

**Comment by wpferguson** [comment]
I tested using image_time.lua.  I changed the mode to `synchronize`, selected 2 images and clicked the calculate button to determine the difference.  I got no result, so I added print statements to the script and got no value for exif_datetime_taken.

My setup was create a fresh clone of master, then apply your PR.  I built and installed, then started for the first time and set up.  Imported some images and started testing.  I did check the database to make sure the datetime_taken column was populated in the images table.

I'll add some print statements to lua/image.c to see if I can figure out what is going on (after I pull your latest commits and try again).

**Comment by wpferguson** [comment]
@phweyland did you change the script code at all?  Still not working on my end.

**Comment by wpferguson** [comment]
@phweyland 

```
dt_datetime_img_to_exif input datetimetaken is 63660622472000000
got  from dt_datetime_img_to_exif
```
`dt_datetime_img_to_exif` keeps returning an empty string.  I'll keep digging deeper...

**Comment by wpferguson** [comment]
```
took the read branch
dt_datetime_img_to_exif input datetimetaken is 63660622472000000
dt_datetime_gtimespan_to_exif 330 - past sdt check
dt_datetime_gtimespan_to_exif 333 - past gts check
dt_datetime_gtimespan_to_exif 337 - past gdt check
dt_datetime_gdatetime_to_exif 216 past exif check
dt_datetime_gdatetime_to_exif 219 sdt after g_date_time_format is (null)
dt_datetime_gdatetime_to_exif 228 didn't copy sdt to exif, returning false
dt_datetime_gtimespan_to_exif 339 - sdt of dt_datetime_gdatetime_to_exif was 
dt_datetime_gtimespan_to_exif 340 - res of dt_datetime_gdatetime_to_exif was FALSE
got  from dt_datetime_img_to_exif
lua stack at exif_datetime_taken_member:331(size 3),
	1:userdata /home/bill/Pictures/dttestdatetime/Seniors/Autumn/_7D_2863.CR2
	2:string exif_datetime_taken
	3:string 
6.522494 LUA exiftime argument is 
```

`g_date_time_format` is returning null, but gdt is not null - any ideas?

**Comment by phweyland** [comment]
> did you change the script code at all?

no, nothing on the script side.

The value GTimeSpan looks OK. 
In metadata_view or in collect, do the date of your images looks OK ?

**Comment by wpferguson** [comment]
> In metadata_view or in collect, do the date of your images looks OK ?

Yes

> In dt_datetime_gdatetime_to_exif() if res is FALSE, the exif date is "".

That worked, but it should not have been FALSE :)

> Looks like your machine doesn't like DT_DATETIME_EXIF_FORMAT. If you replace it by "%Y:%m:%d %H:%M:%S.%f" ?

No, that didn't work either.  In the current `DT_DATETIME_EXIF_FORMAT`, you have colons between %Y, %M, and %D.  Did you mean to use `/`

> What is your locale ?

LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

> In geotagging, does the date look OK ?

geotagging

**Comment by wpferguson** [comment]
OS is Ubuntu 20.04, up to date

**Comment by phweyland** [comment]
geotagging doesn't look OK.

Semi-colon is the Exif standard. 

> That worked, but it should not have been FALSE :)

Right :)
g_date_time_format() fails. Why ? ...

**Comment by phweyland** [comment]
g_date_time_format() returns a string:
"a newly allocated string formatted to the requested format or NULL in the case that there was an error (such as a format specifier not being supported in the current locale)."
The only thing I've added to the former one is ".%f" to get the microsecond.
Could you try to remove it ? (If the milliseconds are not set in lighttable preferences that should not break anything).

**Comment by wpferguson** [comment]
It doesn't like the `.%f` in the format string.  If I take that off, it works fine.  Since we're storing the `datetime_taken` as in integer the floating point isn't really there anyway.  This also fixes geotagging date display

**Comment by wpferguson** [comment]
Now I'll get back to checking the lua stuff :)

**Comment by phweyland** [comment]
Thank you so much for your tests ! At least I have now a way forward to find a workaround !

**Comment by phweyland** [comment]
BTW, what is your libglib2.0-dev version ? Mine is  2.70.4-1.

**Comment by wpferguson** [comment]
Mine is 2.64.6-1

**Comment by phweyland** [comment]
@wpferguson, still work for me. Hopefully for you too now.

**Comment by wpferguson** [comment]
@phweyland SUCCESS!

**Comment by wpferguson** [comment]
@phweyland Lua works fine too.  Thanks

**Comment by TurboGit** [comment]
@phweyland : So I should remove the Draft from here, right?

**Comment by phweyland** [comment]
> So I should remove the Draft from here, right?

Well, At some point, I can read my code again and again, I don't see any other point to solve. Of course a new eye will But yes, I 'm done on my side.

**Comment by wpferguson** [comment]
Looks good to me also.  Lua stuff is fine

**Comment by TurboGit** [comment]
@phweyland : Ok, I expected that you'll be removing the Draft mode when done :) I'll review this and merge if all good.

File: src/common/camera_control.c [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #endif
 #include "common/camera_control.h"
 #include "common/exif.h"
+#include "common/datetime.h"
```
Comment (TurboGit) [inline src/common/camera_control.c:23]: Probably superfluous since nothing else if changed here. I'll test and remove if possible.

**Comment by phweyland** [comment]
> Ok, I expected that you'll be removing the Draft mode when done :

Can I do that ? I didn't know :)

**Review by TurboGit — APPROVED** [review]
Code looks ok to me. I have done a "fast" pass on it as it is quite large. Anyway, I'll test this (probably tomorrow now) and will merge it working on my side. Thanks.

**Review by TurboGit — APPROVED** [review]
Works for me, merging now. Thanks.

File: src/common/camera_control.c (hunk 1 above)
Comment (TurboGit) [reply src/common/camera_control.c:23]: And finally this is needed.

**Pull request merged**